### PR TITLE
[chore] - chore(UX): impacted assistant datasource

### PR DESCRIPTION
**Pull request by JulesBelveze**
## Description

This PR aims at displaying how many assistants use a given data source (website or folder).

This is based on the following task: https://github.com/dust-tt/tasks/issues/849

See screenshots below:

Screenshot 2024-07-02 at 15 26 22

Screenshot 2024-07-02 at 15 25 59

### CODE REVIEW

File: front/lib/api/agent_data_sources.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  }
+  const workspace = await Workspace.findOne({
+    where: { sId: owner.sId },
+  });
+  if (!workspace) {
+    throw new Error(`Workspace not found for sId: ${owner.sId}`);
+  }
+  const wId = workspace.id;
+
+  const dataSourceConfigurations = await AgentDataSourceConfiguration.findAll({
```
Comment (spolu) [inline front/lib/api/agent_data_sources.ts:36]: Name is not ideal let's have agent in the variable name

File: front/lib/api/agent_data_sources.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  if (!owner || !auth.isUser()) {
+    return [];
+  }
+  const workspace = await Workspace.findOne({
+    where: { sId: owner.sId },
+  });
+  if (!workspace) {
+    throw new Error(`Workspace not found for sId: ${owner.sId}`);
+  }
+  const wId = workspace.id;
```
Comment (spolu) [inline front/lib/api/agent_data_sources.ts:34]: No need to pull workspace when you have a owner which has a Model I'd

Let's avoid using wId which generally means workspace sId

Use owner.id directly 👍

File: front/lib/api/agent_data_sources.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+  const dataSourceConfigurations = await AgentDataSourceConfiguration.findAll({
+    include: [
+      {
+        model: DataSource,
+        as: "dataSource",
+        where: {
+          workspaceId: wId,
+          connectorProvider: providerFilter,
```
Comment (spolu) [inline front/lib/api/agent_data_sources.ts:43]: Is this indexed?

File: front/lib/api/agent_data_sources.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
+import type { ConnectorProvider } from "@dust-tt/types";
+
+import type { Authenticator } from "@app/lib/auth";
+import { AgentDataSourceConfiguration } from "@app/lib/models/assistant/actions/data_sources";
+import { DataSource } from "@app/lib/models/data_source";
+import { User } from "@app/lib/models/user";
+import { Workspace } from "@app/lib/models/workspace";
+
+export type AgentEnabledDataSource = {
```
Comment (spolu) [inline front/lib/api/agent_data_sources.ts:9]: This type is kind of hard to understand without context on what we're trying to do here

File: front/lib/api/agent_data_sources.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  dataSourceId: number;
+};
+
+export async function getAgentEnabledDataSources({
```
Comment (spolu) [inline front/lib/api/agent_data_sources.ts:13]: The name of this function is hard to parse as well.

Let's return what we need which is the count by and return a type that is more explicit in that direction 👍

File: front/lib/api/agent_data_sources.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      {
+        model: DataSource,
+        as: "dataSource",
+        where: {
+          workspaceId: wId,
+          connectorProvider: providerFilter,
+        },
+        include: [
+          {
+            model: User,
```
Comment (spolu) [inline front/lib/api/agent_data_sources.ts:47]: Why do you need the user given that you're bit returning it?

**Comment by JulesBelveze** [comment]
@spolu sorry first draft was messy, got a bit confused by two tasks that I tried to tackle at once 🙏🏼

File: front/lib/api/agent_data_sources.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  auth: Authenticator;
+  providerFilter: ConnectorProvider | null;
+}): Promise<DataSourcesUsageByAgent> {
+  const owner = auth.workspace();
+
+  // This condition is critical it checks that we can identify the workspace and that the current
+  // auth is a user for this workspace. Checking `auth.isUser()` is critical as it would otherwise
+  // be possible to access data sources without being authenticated.
+  if (!owner || !auth.isUser()) {
+    return [];
```
Comment (spolu) [inline front/lib/api/agent_data_sources.ts:23]: should be {} no?

File: front/lib/api/agent_data_sources.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+      ],
+      group: ["dataSource.id"],
+      raw: true,
+    });
+  return agentDataSourceConfigurations.reduce((acc, dsConfig) => {
+    acc[dsConfig.dataSourceId] = (
+      dsConfig as unknown as { count: number }
+    ).count;
+    return acc;
+  }, {} as DataSourcesUsageByAgent);
```
Comment (spolu) [inline front/lib/api/agent_data_sources.ts:50]: Any chance we can get rid of this `as` ?

File: front/lib/api/agent_data_sources.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
+import type { ConnectorProvider } from "@dust-tt/types";
+import { Sequelize } from "sequelize";
+
+import type { Authenticator } from "@app/lib/auth";
+import { AgentDataSourceConfiguration } from "@app/lib/models/assistant/actions/data_sources";
+import { DataSource } from "@app/lib/models/data_source";
+
+export type DataSourcesUsageByAgent = Record<number, number>;
```
Comment (spolu) [inline front/lib/api/agent_data_sources.ts:8]: Should be Record<ModelId, number>?

File: front/lib/models/data_source.ts [hunk 10]
```diff
@@ -85,7 +85,10 @@ DataSource.init(
   {
     modelName: "data_source",
     sequelize: frontSequelize,
-    indexes: [{ fields: ["workspaceId", "name"], unique: true }],
+    indexes: [
+      { fields: ["workspaceId", "name"], unique: true },
+      { fields: ["connectorProvider"] },
```
Comment (spolu) [inline front/lib/models/data_source.ts:90]: should be workspaceId, connectorProvider ?

File: front/lib/api/agent_data_sources.ts (hunk 9 above)
Comment (JulesBelveze) [reply front/lib/api/agent_data_sources.ts:8]: Hmmm keys are the `dataSource.id`, from I've seen `ModelId` is for `connectorId`?

File: front/lib/api/agent_data_sources.ts (hunk 8 above)
Comment (JulesBelveze) [reply front/lib/api/agent_data_sources.ts:50]: Yes can do, by setting `reduce<DataSourcesUsageByAgent>`

File: front/lib/api/agent_data_sources.ts (hunk 9 above)
Comment (spolu) [reply front/lib/api/agent_data_sources.ts:8]: It's the type for all models internal IDs 👍

Comment (JulesBelveze) [reply front/lib/api/agent_data_sources.ts]: Alrighty 👍

**Review by spolu — APPROVED** [review]
LGTM
